### DISCUSSION TITLE
BNF path analysis

### DISCUSSION THREAD

**Discussion by mercurial-moon** [networkx/networkx#7605] [Q&A] [answered]
Given a bnf (antlr g4 format) rule like so

`common_table_expression
    : table_name (OPEN_PAR column_name ( COMMA column_name)* CLOSE_PAR)? AS_ OPEN_PAR select_stmt CLOSE_PAR
;`

Railroad diagram of the same rule:

![image](https://github.com/user-attachments/assets/6c1786e6-f645-41f6-a92a-a653fa042549)

I want to find all possible paths for this rule from left to right, so for example the outputs would be

Path1 : `table_name AS_ OPEN_PAR select_stmt CLOSE_PAR`
Path2 : `table_name OPEN_PAR column_name CLOSE_PAR AS_ OPEN_PAR select_stmt CLOSE_PAR`
...
and so on
would this be possible with networkX.
I have already parsed the g4 file into a tree like format.

**Comment by rossbar** [comment, marked as answer]
>  would this be possible with networkX.

Yes - [`all_simple_paths`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.simple_paths.all_simple_paths.html#networkx.algorithms.simple_paths.all_simple_paths) should do the trick.

>  I have already parsed the g4 file into a tree like format.

NetworkX functionality is designed to work on NetworkX graph objects; in this case `nx.DiGraph`. It should be relatively straight-forward to parse your existing data structure into one that `nx.DiGraph` can understand (e.g. including but not limited to an edgelist, dict-of-dict adjacency mapping, adjacency matrix, etc.)

For example, to create the graph above from scratch (i.e. creating the edge list by hand), you could do something like:

```python
>>> import networkx as nx
>>> G = nx.DiGraph([
...     ("table_name", "AS_"),
...     ("table_name", "OPEN_PAR"),
...     ("OPEN_PAR", "column_name"),
... <etc>
```

Of course, if you already have code to parse the original data, you'd probably want to use that as a starting point for creating a networkx graph!
